### ISSUE TITLE
Read of IWebElement raises a CanopyReadException

### ISSUE THREAD

**Issue by milkeg** [lefthandedgoat/canopy#364]
Hello,

Since we updated to Canopy 1.2.2 from 1.0.1, we are facing a CanopyReadException while trying to read lists of IWebElement using the Read function.
This issue is not happening when we run the test cases in debug mode (&&&&)

As we are getting a CanopyReadException, we presumed that the IWebElement is still reachable otherwise we would get a NoSuchElement exception

We check the variable ourList but it seemed properly defined, it contains 10 IWebElements in each list (innerListOne and innerListTwo).

Code:
```
let ourList: IWebElement list list = [elements innerListOne; elements innerListTwo]
// Apply Double.Parse to every elements of the lists within the list ourList
let ourListParsed = ourList |> List.map (List.map (read >> Decimal.Parse))
```
Exception:
```
Error: was unable to read item for unknown reason
Url: ***
Exception details: 
canopy.types+CanopyReadException: was unable to read item for unknown reason
  at canopy.wait.innerwait@10[b] (System.Diagnostics.Stopwatch sw, System.Double timeout, Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] f) [0x00095] in <58f3b205d8107025a745038305b2f358>:0 
  at canopy.wait.waitResults[a] (System.Double timeout, Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] f) [0x00009] in <58f3b205d8107025a745038305b2f358>:0 
  at canopy.core.waitFor2 (System.String message, Microsoft.FSharp.Core.FSharpFunc`2[T,TResult] f) [0x00008] in <58f3b205d8107025a745038305b2f358>:0 
```

When does this exception happen ?

Thank you,
Michael

**Comment by milkeg** [comment]
It's weird because if we call a function instead of using "read >> Double.Parse", I am not facing any exceptions...

```
let foo (elem: IWebElement) =
    let elemRead = read elem
    describe (sprintf "elemRead %s" elemRead)
    let elemParsed = elemRead |> Decimal.Parse
    describe (sprintf "elemParsed %f" elemParsed)
    elemParsed

let ourList: IWebElement list list = [elements innerListOne; elements innerListTwo]
// Apply Double.Parse to every elements of the lists within the list ourList
let ourListParsed = ourList |> List.map (List.map (foo))
```

Thank you,
Michael

**Comment by lefthandedgoat** [comment]
I can't see anything obvious.  The only thing I can think of is that it does not work correctly with function composition `>>`

I will try to reproduce.

**Comment by milkeg** [comment]
Hello,
The issue is still happening even with the workaround we tried :(

We would like to know some more information on how Canopy is behaving.
- Is an IWebElement a kind of "pointer" to a actual HTML element ? Therefore, when we try to read it, Canopy/Selenium is trying go back to the Web page and read that element ?
- How would Canopy/Selenium react if we try to read an IWebElement that has been removed from the Web page ?

Thank you,
Michael

**Comment by lefthandedgoat** [comment]
Yes IWebElement is a representation of an html element from Selenium.  `read` can take two differnt types of arguments, either a string, representing the selector, or an actual element.  Then it will read it.

`read` when passed an element that has been removed from the page or gone stale will not behave very nicely, and is probably what you are seeing.

Here is the source for read: https://github.com/lefthandedgoat/canopy/blob/master/src/canopy/canopy.fs#L391-L410

I didn't realize you were passing IWebElements to read until I re-read your problem.  If you can pass in selectors instead of IWebElements, `read` will try and try again until the element becomes available, and if it does not, it should bubble up the correct error message that the element was not found.

**Comment by milkeg** [comment]
Thank you for this information.

Unfortunately, as we are getting all the IWebElements first, it's difficult to read them at the same time.
```
        let table: (IWebElement * IWebElement) =
            (elements column1, elements column2)
```

As workaround, we implemented a retry mechanism to ensure the element is properly attached to the DOM (and not stale).
```
    let MAX_READ_ATTEMPTS = 1

    let rec ReadAndParseCounted (elem: IWebElement) (count: int): decimal =
        if count > MAX_READ_ATTEMPTS then
            failwith "Not possible to read the element"
        try
            let elemParsed = read elem |> Decimal.Parse
            elemParsed
        with ex ->
            describe "Retrying to read the element..."
            let newCounter = count + 1
            ReadAndParseCounted elem newCounter

    let ReadAndParse (elem: IWebElement): decimal =
        ReadAndParseCounted elem 0
```

Thank you,
Michael

**Comment by lefthandedgoat** [comment]
Ok I am glad you found a work around.  Working with selenium directly can be painful =(, hence why canopy exists.
